examples/msg.c

```diff
@@ -0,0 +1,548 @@
+ * posting a send buffer, if the provider is not ready to process messages,
+ * it may return -FI_EAGAIN. This does not indicate an error, but rather that
+ * the application should try again later. Just like the receive, we drive
+ * progress with fi_cq_read if this is the case.
+ */
+static int post_send(void)
+{
+	char *msg = "Hello, server! I am the client you've been waiting for!\0";
+	int ret;
+
+	(void) snprintf(buffer, BUF_SIZE, "%s", msg);
+
+	do {
+		ret = fi_send(ep, buffer, BUF_SIZE, NULL, fi_addr, NULL);
+		if (ret && ret != -FI_EAGAIN) {
+			printf("error posting send buffer (%d)\n", ret);
+			return ret;
+		}
+		if (ret == -FI_EAGAIN)
+			(void) fi_cq_read(cq, NULL, 0);
+	} while (ret);
+
+	return 0;
+}
+
+/*
+ * Wait for the message to be sent/received using the CQ. fi_cq_read not only
+ * drives progress but also returns any completed events to notify the
+ * application that it can reuse the send/recv buffer. The returned completion
+ * entry will have fields set to let the application know what operation
+ * completed. Not all fields will be valid. The fields set will be indicated
+ * by the cq format (when creating the CQ). In this example, we use
+ * FI_CQ_FORMAT_MSG in order to use the flags field.
+ */
+static int wait_cq(void)
+{
+	struct fi_cq_err_entry comp;
+	int ret;
+
+	do {
+		ret = fi_cq_read(cq, &comp, 1);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			printf("error reading cq (%d)\n", ret);
+			return ret;
+		}
+	} while (ret != 1);
+
+	if (comp.flags & FI_RECV)
+		printf("I received a message!\n");
+	else if (comp.flags & FI_SEND)
+		printf("My message got sent!\n");
+
+	return 0;
+}
+
+static int run(void)
+{
+	int ret;
+
+	if (dst_addr) {
+		printf("Client: send to server %s\n", dst_addr);
+
+		ret = post_send();
+		if (ret)
+			return ret;
+
+		ret = wait_cq();
+		if (ret)
+			return ret;
+
+	} else {
+		printf("Server: post buffer and wait for message from client\n");
+
+		ret = post_recv();
+		if (ret)
+			return ret;
+
+		ret = wait_cq();
+		if (ret)
+			return ret;
+
+		printf("This is the message I received: %s\n", buffer);
+	}
+	return 1;
+}
+
+int main(int argc, char **argv)
+{
+	int ret;
+
+	/*
+	 * Hints are used to request support for specific features
+	 * from a provider
+	 */
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	/*
+	 * Server run with no args, client has server's address
+	 * as an argument
+	 */
+	dst_addr = argv[1];
+
+	/*
+	 * Set anything in hints that the application needs
+	 */
+
+	/*
+	 * Request FI_EP_MSG (reliable datagram) endpoint which will allow us
+	 * to reliably send messages to peers without having
+	 * to listen/connect/accept.
+	 */
+	hints->ep_attr->type = FI_EP_MSG;
+
+	/*
+	 * Request basic messaging capabilities from the provider
+	 * (no tag matching, * no RMA, no atomic operations)
+	 */
+	hints->caps = FI_MSG;
+
+	/* Specifically request the tcp provider for the simple test */
+	hints->fabric_attr->prov_name = "tcp";
```

Comment:
With the provider explicitly set to `tcp` here, the use of the `FI_PROVIDER` environment makes no sense. Either change here or the usage in the README file.